File: grafana/tabs/monit.py
"""
Monit tab — Monit service/resource monitoring metrics (opnsense_monit_*).

Plugin-gated: tab hidden unless monit metrics are present.

All metrics are gauges (instantaneous values, no _total counters).
"""

from builder import Builder, sel, RUNSTOP, OKERR, UPDOWN, MONITORED


def build(b: Builder):
    b.sentinel("has_monit", metric="opnsense_monit_status_ok")

    # ------------------------------------------------------------------ #
    # Row 1: Monit Overview                                                #
    # ------------------------------------------------------------------ #
    svc = b.stat(
        "Monit Service",
        sel("opnsense_monit_service_running"),
        mappings=RUNSTOP, color_mode="background",
        thresholds=[{"color": "red", "value": None}, {"color": "green", "value": 1}],
        w=4, h=4,
        desc="Monit service state (1 = running, 0 = stopped).",
    )
    status_ok = b.stat(
        "Monit Status Reachable",
        sel("opnsense_monit_status_ok"),
        mappings=OKERR, color_mode="background",
        thresholds=[{"color": "red", "value": None}, {"color": "green", "value": 1}],
        w=4, h=4,
        desc=(
            "1 when the monit httpd is reachable and returned a valid status XML. "
            "0 when monit is unreachable or returned an error payload."
        ),
    )
    checks_total = b.stat(
        "Checks Configured",
        sel("opnsense_monit_checks_total"),
        unit="short", w=4, h=4,
        desc="Total number of configured monit checks.",
    )
    checks_ok = b.stat(
        "Checks OK",
        f'sum({sel("opnsense_monit_check_status")})',
        unit="short", w=4, h=4,
        thresholds=[
            {"color": "red", "value": None},
            {"color": "green", "value": 1},
        ],
        desc=("Number of monit checks currently in OK state (status field == 0)."
             "Fleet total: this is a deliberate sum across every selected firewall (#468) — with two boxes picked, the number is both boxes' together."),
    )
    checks_monitored = b.stat(
        "Checks Monitored",
        f'sum({sel("opnsense_monit_check_monitored")})',
        unit="short", w=4, h=4,
        desc=("Number of monit checks actively being monitored (monitor != 0)."
             "Fleet total: this is a deliberate sum across every selected firewall (#468) — with two boxes picked, the number is both boxes' together."),
    )

    # ------------------------------------------------------------------ #
    # Row 2: Check Status Detail                                           #
    # ------------------------------------------------------------------ #
    check_status_timeline = b.statetimeline(
        "Check Status Over Time",
        [(sel("opnsense_monit_check_status"), "{{name}} ({{type}})")],
        UPDOWN, w=24, h=8,
        desc=(
            "Monit check status over time per check. "
            "1 = OK (monit status field == 0), 0 = failed/error."
        ),
    )
    check_monitored_timeline = b.statetimeline(
        "Check Monitored State",
        [(sel("opnsense_monit_check_monitored"), "{{name}} ({{type}})")],
        MONITORED, w=24, h=8,
        desc=(
            "Whether each check is actively monitored over time. "
            "1 = monitored (monitor != 0), 0 = unmonitored."
        ),
    )
    checks_table = b.table(
        "Check Details",
        [
            sel("opnsense_monit_check_status"),
            sel("opnsense_monit_check_monitored"),
        ],
        w=24, h=10,
        excludes=["__name__", "job", "instance"],
        renames={
            "name": "Check Name",
            "type": "Type",
            "Value #A": "Status OK",
            "Value #B": "Monitored",
        },
        sort_by="Check Name",
        desc="Current status and monitored state for all configured monit checks.",
    )

    # ------------------------------------------------------------------ #
    # Row 3: Filesystem Check Resources (#219)                             #
    # ------------------------------------------------------------------ #
    fs_usage_ts = b.ts(
        "Filesystem Usage",
        [
            (sel("opnsense_monit_filesystem_usage_percent"), "{{name}} block"),
            (sel("opnsense_monit_filesystem_inode_usage_percent"), "{{name}} inode"),
        ],
        unit="percent", w=24, h=8,
        desc=(
            "opnsense_monit_filesystem_usage_percent / "
            "opnsense_monit_filesystem_inode_usage_percent: block (space) and inode usage "
            "percent for each monit filesystem check. The only disk-usage-percent source in "
            "the exporter besides the health subsystem gauge."
        ),
    )

    # ------------------------------------------------------------------ #
    # Row 4: Process Check Resources (#219)                                #
    # ------------------------------------------------------------------ #
    process_cpu_ts = b.ts(
        "Process CPU Usage",
        [(sel("opnsense_monit_process_cpu_percent"), "{{name}}")],
        unit="percent", w=12, h=8,
        desc=(
            "opnsense_monit_process_cpu_percent: CPU usage percent for each monit process "
            "check. Absent until monit's second poll cycle computes a rate (monit reports -1 "
            "beforehand; the exporter treats that sentinel as no data)."
        ),
    )
    process_mem_ts = b.ts(
        "Process Memory Usage",
        [(sel("opnsense_monit_process_memory_bytes"), "{{name}}")],
        unit="bytes", w=12, h=8,
        desc="opnsense_monit_process_memory_bytes: resident memory usage for each monit process check.",
    )
    process_uptime_ts = b.ts(
        "Process Uptime",
        [(sel("opnsense_monit_process_uptime_seconds"), "{{name}}")],
        unit="s", w=12, h=8,
        desc="opnsense_monit_process_uptime_seconds: uptime for each monit process check.",
    )
    process_threads_ts = b.ts(
        "Process Threads",
        [(sel("opnsense_monit_process_threads"), "{{name}}")],
        unit="short", w=12, h=8,
        desc="opnsense_monit_process_threads: thread count for each monit process check.",
    )

    # ------------------------------------------------------------------ #
    # Row 5: Host Check Resources (#219)                                   #
    # ------------------------------------------------------------------ #
    host_icmp_ts = b.ts(
        "Host ICMP Response Time",
        [(sel("opnsense_monit_host_icmp_response_seconds"), "{{name}}")],
        unit="s", w=12, h=8,
        desc="opnsense_monit_host_icmp_response_seconds: ICMP ping response time for each monit host check.",
    )
    host_port_ts = b.ts(
        "Host Port Response Time",
        [(sel("opnsense_monit_port_response_seconds"), "{{name}} :{{port}} ({{protocol}})")],
        unit="s", w=12, h=8,
        desc=(
            "opnsense_monit_port_response_seconds: connection response time for each monit "
            "host check's port/connection test."
        ),
    )

    # ------------------------------------------------------------------ #
    # Row 6: System Check Resources (#219)                                 #
    # ------------------------------------------------------------------ #
    system_load_ts = b.ts(
        "System Check Load Average",
        [(sel("opnsense_monit_system_load"), "{{name}} {{period}}")],
        unit="short", w=12, h=8,
        desc="opnsense_monit_system_load: load average (period=1m/5m/15m) for each monit system check.",
    )
    system_cpu_ts = b.ts(
        "System Check CPU Usage",
        [(sel("opnsense_monit_system_cpu_percent"), "{{name}} {{mode}}")],
        unit="percent", w=12, h=8,
        desc=(
            "opnsense_monit_system_cpu_percent: CPU usage percent by mode for each monit "
            "system check. Reported modes are platform-dependent (OPNsense/FreeBSD reports "
            "user/system/nice/hardirq)."
        ),
    )
    system_mem_swap_ts = b.ts(
        "System Check Memory & Swap",
        [
            (sel("opnsense_monit_system_memory_percent"), "{{name}} memory"),
            (sel("opnsense_monit_system_swap_percent"), "{{name}} swap"),
        ],
        unit="percent", w=24, h=8,
        desc=(
            "opnsense_monit_system_memory_percent / opnsense_monit_system_swap_percent: "
            "memory and swap usage percent for each monit system check."
        ),
    )

    b.tab("Monit", [
        b.row("Monit Overview",
              [svc, status_ok, checks_total, checks_ok, checks_monitored],
              present="has_monit"),
        b.row("Check Status Detail",
              [check_status_timeline, check_monitored_timeline, checks_table],
              present="has_monit"),
        b.row("Filesystem Check Resources",
              [fs_usage_ts],
              present="has_monit"),
        b.row("Process Check Resources",
              [process_cpu_ts, process_mem_ts, process_uptime_ts, process_threads_ts],
              present="has_monit"),
        b.row("Host Check Resources",
              [host_icmp_ts, host_port_ts],
              present="has_monit"),
        b.row("System Check Resources",
              [system_load_ts, system_cpu_ts, system_mem_swap_ts],
              present="has_monit"),
    ])

File: internal/collector/monit.go
package collector

import (
	"context"
	"log/slog"

	"github.com/prometheus/client_golang/prometheus"
	"github.com/rknightion/opnsense-exporter/opnsense"
)

// monitCollector collects monit service-check metrics from the OPNsense
// Monit API (api/monit/status/get/xml).
type monitCollector struct {
	log *slog.Logger

	serviceRunning *prometheus.Desc
	statusOK       *prometheus.Desc
	checksTotal    *prometheus.Desc
	checkStatus    *prometheus.Desc
	checkMonitored *prometheus.Desc

	// Per-check resource telemetry (#219). All gauges; a value is only
	// emitted when the underlying monit payload actually carried it — never
	// coerced to zero when absent.
	filesystemUsagePercent      *prometheus.Desc
	filesystemInodeUsagePercent *prometheus.Desc
	processCPUPercent           *prometheus.Desc
	processMemoryBytes          *prometheus.Desc
	processUptimeSeconds        *prometheus.Desc
	processThreads              *prometheus.Desc
	hostICMPResponseSeconds     *prometheus.Desc
	portResponseSeconds         *prometheus.Desc
	systemLoad                  *prometheus.Desc
	systemMemoryPercent         *prometheus.Desc
	systemSwapPercent           *prometheus.Desc
	systemCPUPercent            *prometheus.Desc

	subsystem string
	instance  string
}

func init() {
	collectorInstances = append(collectorInstances, &monitCollector{
		subsystem: MonitSubsystem,
	})
}

func (c *monitCollector) Name() string { return c.subsystem }

func (c *monitCollector) Register(namespace, instanceLabel string, log *slog.Logger) {
	c.log = log
	c.instance = instanceLabel
	c.log.Debug("Registering collector", "collector", c.Name())

	c.serviceRunning = buildPrometheusDesc(c.subsystem, "service_running",
		"Whether the Monit service is running (1 = running, 0 = stopped/disabled)", nil)

	c.statusOK = buildPrometheusDesc(c.subsystem, "status_ok",
		"Whether the monit httpd was reachable and returned a valid status (1 = ok, 0 = failed/unreachable)", nil)

	c.checksTotal = buildPrometheusDesc(c.subsystem, "checks_total",
		"Total number of service checks configured in monit", nil)

	c.checkStatus = buildPrometheusDesc(c.subsystem, "check_status",
		"Whether a monit check reports no errors (1 = status field is 0, 0 = error)",
		[]string{"name", "type"})

	c.checkMonitored = buildPrometheusDesc(c.subsystem, "check_monitored",
		"Whether a monit check is actively monitored (1 = monitored, 0 = not monitored)",
		[]string{"name", "type"})

	c.filesystemUsagePercent = buildPrometheusDesc(c.subsystem, "filesystem_usage_percent",
		"Filesystem block (space) usage percent for a monit filesystem check",
		[]string{"name", "type"})

	c.filesystemInodeUsagePercent = buildPrometheusDesc(c.subsystem, "filesystem_inode_usage_percent",
		"Filesystem inode usage percent for a monit filesystem check",
		[]string{"name", "type"})

	c.processCPUPercent = buildPrometheusDesc(c.subsystem, "process_cpu_percent",
		"CPU usage percent for a monit process check (absent until monit's second poll cycle computes a rate)",
		[]string{"name", "type"})

	c.processMemoryBytes = buildPrometheusDesc(c.subsystem, "process_memory_bytes",
		"Resident memory usage in bytes for a monit process check",
		[]string{"name", "type"})

	c.processUptimeSeconds = buildPrometheusDesc(c.subsystem, "process_uptime_seconds",
		"Uptime in seconds for a monit process check",
		[]string{"name", "type"})

	c.processThreads = buildPrometheusDesc(c.subsystem, "process_threads",
		"Number of threads for a monit process check",
		[]string{"name", "type"})

	c.hostICMPResponseSeconds = buildPrometheusDesc(c.subsystem, "host_icmp_response_seconds",
		"ICMP ping response time in seconds for a monit host check",
		[]string{"name", "type"})

	c.portResponseSeconds = buildPrometheusDesc(c.subsystem, "port_response_seconds",
		"Port connection response time in seconds for a monit host check's port test",
		[]string{"name", "type", "port", "protocol"})

	c.systemLoad = buildPrometheusDesc(c.subsystem, "system_load",
		"System load average for a monit system check",
		[]string{"name", "type", "period"})

	c.systemMemoryPercent = buildPrometheusDesc(c.subsystem, "system_memory_percent",
		"System memory usage percent for a monit system check",
		[]string{"name", "type"})

	c.systemSwapPercent = buildPrometheusDesc(c.subsystem, "system_swap_percent",
		"System swap usage percent for a monit system check",
		[]string{"name", "type"})

	c.systemCPUPercent = buildPrometheusDesc(c.subsystem, "system_cpu_percent",
		"System CPU usage percent by mode for a monit system check",
		[]string{"name", "type", "mode"})
}

func (c *monitCollector) Describe(ch chan<- *prometheus.Desc) {
	ch <- c.serviceRunning
	ch <- c.statusOK
	ch <- c.checksTotal
	ch <- c.checkStatus
	ch <- c.checkMonitored
	ch <- c.filesystemUsagePercent
	ch <- c.filesystemInodeUsagePercent
	ch <- c.processCPUPercent
	ch <- c.processMemoryBytes
	ch <- c.processUptimeSeconds
	ch <- c.processThreads
	ch <- c.hostICMPResponseSeconds
	ch <- c.portResponseSeconds
	ch <- c.systemLoad
	ch <- c.systemMemoryPercent
	ch <- c.systemSwapPercent
	ch <- c.systemCPUPercent
}

func (c *monitCollector) Update(ctx context.Context, client *opnsense.Client, ch chan<- prometheus.Metric) *opnsense.APICallError {
	// Fetch the Monit service status (running/stopped/disabled) first. This
	// uses FetchServiceStatusOptional so a missing endpoint is silent — a 404
	// from the service endpoint means the plugin is absent and we emit nothing.
	svcStatus, svcPresent, sErr := client.FetchServiceStatusOptional("monitServiceStatus")
	if sErr != nil {
		c.log.Warn("failed to fetch monit service status", "err", sErr)
		// Do not return early — still attempt to fetch monit status data.
	} else if svcPresent {
		running := 0.0
		if svcStatus == "running" {
			running = 1.0
		}
		ch <- prometheus.MustNewConstMetric(c.serviceRunning, prometheus.GaugeValue,
			running, c.instance)
	}

	// Fetch the monit status (check data via monit's httpd).
	data, err := client.FetchMonitStatus()
	if err != nil {
		return err
	}

	// Emit status_ok regardless of StatusOK value so the absence of monit
	// connectivity is always visible in the metric.
	statusOKVal := 0.0
	if data.StatusOK {
		statusOKVal = 1.0
	}
	ch <- prometheus.MustNewConstMetric(c.statusOK, prometheus.GaugeValue,
		statusOKVal, c.instance)

	// Only emit per-check metrics when monit is reachable and returned data.
	if !data.StatusOK {
		return nil
	}

	ch <- prometheus.MustNewConstMetric(c.checksTotal, prometheus.GaugeValue,
		float64(len(data.Checks)), c.instance)

	for _, check := range data.Checks {
		ch <- prometheus.MustNewConstMetric(c.checkStatus, prometheus.GaugeValue,
			check.StatusOK, check.Name, check.Type, c.instance)
		ch <- prometheus.MustNewConstMetric(c.checkMonitored, prometheus.GaugeValue,
			check.Monitored, check.Name, check.Type, c.instance)

		if v := check.FilesystemUsagePercent; v != nil {
			ch <- prometheus.MustNewConstMetric(c.filesystemUsagePercent, prometheus.GaugeValue,
				*v, check.Name, check.Type, c.instance)
		}
		if v := check.FilesystemInodeUsagePercent; v != nil {
			ch <- prometheus.MustNewConstMetric(c.filesystemInodeUsagePercent, prometheus.GaugeValue,
				*v, check.Name, check.Type, c.instance)
		}

		if v := check.ProcessCPUPercent; v != nil {
			ch <- prometheus.MustNewConstMetric(c.processCPUPercent, prometheus.GaugeValue,
				*v, check.Name, check.Type, c.instance)
		}
		if v := check.ProcessMemoryBytes; v != nil {
			ch <- prometheus.MustNewConstMetric(c.processMemoryBytes, prometheus.GaugeValue,
				*v, check.Name, check.Type, c.instance)
		}
		if v := check.ProcessUptimeSeconds; v != nil {
			ch <- prometheus.MustNewConstMetric(c.processUptimeSeconds, prometheus.GaugeValue,
				*v, check.Name, check.Type, c.instance)
		}
		if v := check.ProcessThreads; v != nil {
			ch <- prometheus.MustNewConstMetric(c.processThreads, prometheus.GaugeValue,
				*v, check.Name, check.Type, c.instance)
		}

		if v := check.ICMPResponseSeconds; v != nil {
			ch <- prometheus.MustNewConstMetric(c.hostICMPResponseSeconds, prometheus.GaugeValue,
				*v, check.Name, check.Type, c.instance)
		}
		for _, port := range check.Ports {
			ch <- prometheus.MustNewConstMetric(c.portResponseSeconds, prometheus.GaugeValue,
				port.ResponseSeconds, check.Name, check.Type, port.Port, port.Protocol, c.instance)
		}

		if v := check.SystemLoad1; v != nil {
			ch <- prometheus.MustNewConstMetric(c.systemLoad, prometheus.GaugeValue,
				*v, check.Name, check.Type, "1m", c.instance)
		}
		if v := check.SystemLoad5; v != nil {
			ch <- prometheus.MustNewConstMetric(c.systemLoad, prometheus.GaugeValue,
				*v, check.Name, check.Type, "5m", c.instance)
		}
		if v := check.SystemLoad15; v != nil {
			ch <- prometheus.MustNewConstMetric(c.systemLoad, prometheus.GaugeValue,
				*v, check.Name, check.Type, "15m", c.instance)
		}
		if v := check.SystemMemoryPercent; v != nil {
			ch <- prometheus.MustNewConstMetric(c.systemMemoryPercent, prometheus.GaugeValue,
				*v, check.Name, check.Type, c.instance)
		}
		if v := check.SystemSwapPercent; v != nil {
			ch <- prometheus.MustNewConstMetric(c.systemSwapPercent, prometheus.GaugeValue,
				*v, check.Name, check.Type, c.instance)
		}
		for mode, v := range check.SystemCPU {
			ch <- prometheus.MustNewConstMetric(c.systemCPUPercent, prometheus.GaugeValue,
				v, check.Name, check.Type, mode, c.instance)
		}
	}

	return nil
}
